tooling/hibernate-gradle-plugin/src/main/groovy/org/hibernate/orm/tooling/gradle/EnhancementHelper.java

```diff
@@ -0,0 +1,178 @@
+import org.hibernate.bytecode.enhance.spi.EnhancementContext;
+import org.hibernate.bytecode.enhance.spi.Enhancer;
+import org.hibernate.bytecode.enhance.spi.UnloadedClass;
+import org.hibernate.bytecode.enhance.spi.UnloadedField;
+import org.hibernate.cfg.Environment;
+
+/**
+ * @author Steve Ebersole
+ */
+public class EnhancementHelper {
+	static void enhance(SourceSet sourceSet, EnhanceExtension options, Project project) {
+		final ClassLoader classLoader = toClassLoader( sourceSet.getRuntimeClasspath() );
+
+		final EnhancementContext enhancementContext = new DefaultEnhancementContext() {
+			@Override
+			public ClassLoader getLoadingClassLoader() {
+				return classLoader;
+			}
+
+			@Override
+			public boolean doBiDirectionalAssociationManagement(UnloadedField field) {
+				return options.getEnableAssociationManagement();
+			}
+
+			@Override
+			public boolean doDirtyCheckingInline(UnloadedClass classDescriptor) {
+				return options.getEnableDirtyTracking();
+			}
+
+			@Override
+			public boolean hasLazyLoadableAttributes(UnloadedClass classDescriptor) {
+				return options.getEnableLazyInitialization();
+			}
+
+			@Override
+			public boolean isLazyLoadable(UnloadedField field) {
+				return options.getEnableLazyInitialization();
+			}
+
+			@Override
+			public boolean doExtendedEnhancement(UnloadedClass classDescriptor) {
+				return options.getEnableExtendedEnhancement();
+			}
+		};
+
+		if ( options.getEnableExtendedEnhancement() ) {
+			project.getLogger().warn("Extended enhancement is enabled. Classes other than entities may be modified. You should consider access the entities using getter/setter methods and disable this property. Use at your own risk." );
+		}
+
+		final Enhancer enhancer = Environment.getBytecodeProvider().getEnhancer( enhancementContext );
+
+		for ( File classesDir: sourceSet.getOutput().getClassesDirs() ) {
+			final FileTree fileTree = project.fileTree( classesDir );
+			for ( File file : fileTree ) {
+				if ( !file.getName().endsWith( ".class" ) ) {
+					continue;
+				}
+
+				final byte[] enhancedBytecode = doEnhancement( classesDir, file, enhancer );
+				if ( enhancedBytecode != null ) {
+					writeOutEnhancedClass( enhancedBytecode, file, project.getLogger() );
+					project.getLogger().info( "Successfully enhanced class [" + file + "]" );
+				}
+				else {
+					project.getLogger().info( "Skipping class [" + file.getAbsolutePath() + "], not an entity nor embeddable" );
+				}
+			}
+		}
+	}
+
+	public static ClassLoader toClassLoader(FileCollection runtimeClasspath) {
+		List<URL> urls = new ArrayList<>();
+		for ( File file : runtimeClasspath ) {
+			try {
+				urls.add( file.toURI().toURL() );
+			}
+			catch (MalformedURLException e) {
+				throw new GradleException( "Unable to resolve classpath entry to URL : " + file.getAbsolutePath(), e );
+			}
+		}
+		return new URLClassLoader( urls.toArray( new URL[0] ), Enhancer.class.getClassLoader() );
```

Comment:
The old line in ```HibernatePlugin``` was:

```		return new URLClassLoader( urls.toArray( new URL[urls.size()] ), Enhancer.class.getClassLoader() );```

Was  ```new URL[urls.size()]``` supposed to be changed to ```urls.toArray( new URL[0] )```?

Reply:
Building an array of size 0 is the new recommended way to do this.  The JVM can already dynamically resize an array very efficiently.

Reply:
OK, good to know.